### PR TITLE
Make llvm:Function behave as a value

### LINKED ISSUE #557: Make llvm:FunctionDecl be an llvm:ConstPointerValue
At the moment we have a toValue method to work around that, but we should find a better way.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Make llvm:Function behave as a value
Resolves #557

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by heshanpadmasiri** [comment]
@jclark wouldn't this (making value a union of "old" value (now RegisterValue) and Function types) be a good solution for our problem given that

- For everything outside of llvm this will have no effect. They can use Values as they used to, and also treat functions as Values when needed
- Inside the llvm side there is little to no change required. This is because once we add separate `operand` and `ty` fields (common with all other values) in addition to `function name` and `function type` (common among functions) everything that expect a "old" value are using operand and ty where as Function uses function fields (function name and function type, note that these don't take the same values)

I am asking this because we have bunch of type casts that try to cast RegisterValues ("old" Value) to its sub types and and changes to type of RegisterValues seems to create issues to them.

File: compiler/modules/print.llvm/llvm.bal [hunk 1]
```diff
@@ -11,8 +11,10 @@ import ballerina/io;
 // Operand of an unnamed variable/basic block
 type Unnamed int;
 
-# Corresponds to LLVMValueRef 
-public readonly distinct class Value {
+public type Value RegisterValue|Function;
+
+# Corresponds to LLVMValueRef
+public readonly distinct class RegisterValue {
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: I don't think RegisterValue is a good name. How about DataValue?

File: compiler/modules/print.llvm/tests/testBuilder.bal [hunk 2]
```diff
@@ -8,7 +8,7 @@ function builderStoreTypeCheck() {
     BasicBlock bb = f.appendBasicBlock();
     Builder builder = context.createBuilder();
     builder.positionAtEnd(bb);
-    Value s = new ("i64", "s");
+    Value s = new RegisterValue("i64", "s");
```
Comment (jclark) [inline compiler/modules/print.llvm/tests/testBuilder.bal]: Can you declare s as type RegisterValue?

Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/tests/testBuilder.bal]: fixed

File: compiler/modules/print.llvm/llvm.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: fixed

**Review by jclark — APPROVED** [review]
Looks good. Do we have a test of calling a function ptr that was loaded from memory (e.g. from a struct member)? If not, please add one.

**Pull request merged**